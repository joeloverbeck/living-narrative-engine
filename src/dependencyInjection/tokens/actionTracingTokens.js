import { freeze } from '../../utils/cloneUtils.js';

/**
 * @file Action tracing DI tokens.
 * @typedef {string} DiToken
 */

/**
 * Action tracing tokens for dependency injection.
 *
 * @type {Readonly<Record<string, DiToken>>}
 */
export const actionTracingTokens = freeze({
  IActionTraceConfigLoader: 'IActionTraceConfigLoader',
  IActionTraceConfigValidator: 'IActionTraceConfigValidator',
  IActionTraceFilter: 'IActionTraceFilter',
  IActionExecutionTraceFactory: 'IActionExecutionTraceFactory',
  IActionTraceOutputService: 'IActionTraceOutputService',
  ITraceDirectoryManager: 'ITraceDirectoryManager',
  IActionAwareStructuredTrace: 'IActionAwareStructuredTrace',
  // Event dispatch tracing tokens
  IEventDispatchTracer: 'IEventDispatchTracer',
  // Storage adapter for IndexedDB
  IIndexedDBStorageAdapter: 'IIndexedDBStorageAdapter',
  // JSON formatter for trace output
  IJsonTraceFormatter: 'IJsonTraceFormatter',
  // Human-readable formatter for trace output
  IHumanReadableFormatter: 'IHumanReadableFormatter',
});
