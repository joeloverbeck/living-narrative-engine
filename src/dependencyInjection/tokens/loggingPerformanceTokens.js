/**
 * @file Dependency injection tokens for logging performance monitoring
 * @see ../../logging/loggingPerformanceMonitor.js
 * @see ../../logging/loggingPerformanceReporter.js
 * @see ../../logging/loggingResourceMonitor.js
 * @see ../../logging/loggingPerformanceAdvisor.js
 */

import { freeze } from '../../utils/cloneUtils.js';

/**
 * @typedef {string} DiToken
 */

/**
 * Logging performance monitoring tokens
 * @type {Readonly<Record<string, DiToken>>}
 */
export const loggingPerformanceTokens = freeze({
  ILoggingPerformanceMonitor: 'ILoggingPerformanceMonitor',
  ILoggingPerformanceReporter: 'ILoggingPerformanceReporter',
  ILoggingResourceMonitor: 'ILoggingResourceMonitor',
  ILoggingPerformanceAdvisor: 'ILoggingPerformanceAdvisor',
});

export default loggingPerformanceTokens;
