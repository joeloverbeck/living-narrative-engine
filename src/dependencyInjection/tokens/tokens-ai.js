import { freeze } from '../../utils';

/**
 * @file AI-specific DI tokens.
 * @typedef {string} DiToken
 */

/**
 * Tokens used by the AI subsystems.
 *
 * @type {Readonly<Record<string, DiToken>>}
 */
export const aiTokens = freeze({
  LLMAdapter: 'LLMAdapter',
  ILLMChooser: 'ILLMChooser',
  IActionIndexer: 'IActionIndexer',
  ITurnActionFactory: 'ITurnActionFactory',
  ITurnStateFactory: 'ITurnStateFactory',
  TurnStrategyFactory: 'TurnStrategyFactory',
  AIStrategyFactory: 'AIStrategyFactory',
  ITurnContextFactory: 'ITurnContextFactory',
  HumanStrategyFactory: 'HumanStrategyFactory',
  IPromptBuilder: 'IPromptBuilder',
  IAIGameStateProvider: 'IAIGameStateProvider',
  IAIPromptContentProvider: 'IAIPromptContentProvider',
  IAIPromptPipeline: 'IAIPromptPipeline',
  IHttpClient: 'IHttpClient',
  ILLMResponseProcessor: 'ILLMResponseProcessor',
  IAIFallbackActionFactory: 'IAIFallbackActionFactory',
  PromptBuilder: 'PromptBuilder',
  AIGameStateProvider: 'AIGameStateProvider',
  AIPromptContentProvider: 'AIPromptContentProvider',
  LLMResponseProcessor: 'LLMResponseProcessor',
  ActionIndexingService: 'ActionIndexingService',
  IConfigurationProvider: 'IConfigurationProvider',
  LlmConfigManager: 'LlmConfigManager',
  PlaceholderResolver: 'PlaceholderResolver',
  ExecutionPlaceholderResolver: 'ExecutionPlaceholderResolver',
  StandardElementAssembler: 'StandardElementAssembler',
  PerceptionLogAssembler: 'PerceptionLogAssembler',
  ThoughtsSectionAssembler: 'ThoughtsSectionAssembler',
  NotesSectionAssembler: 'NotesSectionAssembler',
  GoalsSectionAssembler: 'GoalsSectionAssembler',
  IndexedChoicesAssembler: 'IndexedChoicesAssembler',
  IPromptStaticContentService: 'IPromptStaticContentService',
  IPerceptionLogFormatter: 'IPerceptionLogFormatter',
  IGameStateValidationServiceForPrompting:
    'IGameStateValidationServiceForPrompting',
  AssemblerRegistry: 'AssemblerRegistry',
  IEntitySummaryProvider: 'IEntitySummaryProvider',
  IActorStateProvider: 'IActorStateProvider',
  IPerceptionLogProvider: 'IPerceptionLogProvider',
  IAvailableActionsProvider: 'IAvailableActionsProvider',
  ILocationSummaryProvider: 'ILocationSummaryProvider',
  IActorDataExtractor: 'IActorDataExtractor',
  ILLMDecisionProvider: 'ILLMDecisionProvider',
  LlmJsonService: 'LlmJsonService',
});
