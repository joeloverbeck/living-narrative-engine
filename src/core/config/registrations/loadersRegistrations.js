/**
 * @fileoverview Registers data loading services and their core dependencies.
 */

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../../appContainer.js').default} AppContainer */
/** @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger */
/** @typedef {import('../../interfaces/coreServices.js').IConfiguration} IConfiguration */
/** @typedef {import('../../interfaces/coreServices.js').IPathResolver} IPathResolver */
/** @typedef {import('../../interfaces/coreServices.js').ISchemaValidator} ISchemaValidator */
/** @typedef {import('../../interfaces/coreServices.js').IDataRegistry} IDataRegistry */
/** @typedef {import('../../interfaces/coreServices.js').IDataFetcher} IDataFetcher */
/** @typedef {import('../../services/schemaLoader.js').default} SchemaLoader */
/** @typedef {import('../../services/manifestLoader.js').default} ManifestLoader */
/** @typedef {import('../../services/ruleLoader.js').default} RuleLoader */
/** @typedef {import('../../services/genericContentLoader.js').default} GenericContentLoader */
/** @typedef {import('../../services/componentDefinitionLoader.js').default} ComponentDefinitionLoader */
/** @typedef {import('../../services/gameConfigLoader.js').default} GameConfigLoader */ // <<< ADDED
/** @typedef {import('../../services/modManifestLoader.js').default} ModManifestLoader */ // <<< ADDED: MODLOADER-005 A
/** @typedef {import('../../services/staticConfiguration.js').default} StaticConfiguration */
/** @typedef {import('../../services/defaultPathResolver.js').default} DefaultPathResolver */
/** @typedef {import('../../services/ajvSchemaValidator.js').default} AjvSchemaValidator */
/** @typedef {import('../../services/inMemoryDataRegistry.js').default} InMemoryDataRegistry */
/** @typedef {import('../../services/workspaceDataFetcher.js').default} WorkspaceDataFetcher */


// --- Core Service Imports ---
import StaticConfiguration from '../../services/staticConfiguration.js';
import DefaultPathResolver from '../../services/defaultPathResolver.js';
import AjvSchemaValidator from '../../services/ajvSchemaValidator.js';
import InMemoryDataRegistry from '../../services/inMemoryDataRegistry.js';
import WorkspaceDataFetcher from '../../services/workspaceDataFetcher.js';

// --- Loader Imports ---
import SchemaLoader from '../../services/schemaLoader.js';
import ManifestLoader from '../../services/manifestLoader.js';
import RuleLoader from '../../services/ruleLoader.js';
import GenericContentLoader from '../../services/genericContentLoader.js';
import ComponentDefinitionLoader from '../../services/componentDefinitionLoader.js';
import GameConfigLoader from '../../services/gameConfigLoader.js'; // <<< ADDED
import ModManifestLoader from '../../services/modManifestLoader.js'; // <<< ADDED: MODLOADER-005 A


// --- DI & Helper Imports ---
import {tokens} from '../../tokens.js';
import {Registrar} from '../../dependencyInjection/registrarHelpers.js';

/**
 * Registers core data infrastructure services (Configuration, PathResolver, Validator, Registry, Fetcher)
 * and specific data loaders (Schema, Manifest, Rules, Generic Content, Component Definitions, Game Config, Mod Manifests). // <<< UPDATED Description
 *
 * @export
 * @param {AppContainer} container - The application's DI container.
 */
export function registerLoaders(container) {
    const registrar = new Registrar(container);
    // Resolve logger early for use within this registration bundle
    /** @type {ILogger} */
    const logger = container.resolve(tokens.ILogger);
    logger.debug('Loaders Registration: Starting core services and data loaders...');

    // === Core Infrastructure Interfaces ===
    // These are fundamental services needed by the loaders and other parts of the app.
    registrar.singletonFactory(tokens.IConfiguration, () => new StaticConfiguration());
    logger.debug(`Loaders Registration: Registered ${tokens.IConfiguration}.`);

    // DefaultPathResolver depends on IConfiguration
    registrar.singletonFactory(tokens.IPathResolver, (c) => new DefaultPathResolver(c.resolve(tokens.IConfiguration)));
    logger.debug(`Loaders Registration: Registered ${tokens.IPathResolver}.`);

    registrar.singletonFactory(tokens.ISchemaValidator, () => new AjvSchemaValidator());
    logger.debug(`Loaders Registration: Registered ${tokens.ISchemaValidator}.`);

    registrar.singletonFactory(tokens.IDataRegistry, () => new InMemoryDataRegistry());
    logger.debug(`Loaders Registration: Registered ${tokens.IDataRegistry}.`);

    registrar.singletonFactory(tokens.IDataFetcher, () => new WorkspaceDataFetcher());
    logger.debug(`Loaders Registration: Registered ${tokens.IDataFetcher}.`);

    // === Data Loaders ===
    // These services are responsible for fetching, validating, and potentially processing
    // specific types of game data files.

    // SchemaLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, ILogger
    registrar.singletonFactory(tokens.SchemaLoader, (c) => new SchemaLoader(
        c.resolve(tokens.IConfiguration),
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.SchemaLoader}.`);

    // ManifestLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, ILogger
    // NOTE: This might become deprecated or change with modding system.
    registrar.singletonFactory(tokens.ManifestLoader, (c) => new ManifestLoader(
        c.resolve(tokens.IConfiguration),
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.ManifestLoader}.`);

    // RuleLoader depends on IPathResolver, IDataFetcher, ISchemaValidator, IDataRegistry, ILogger
    registrar.singletonFactory(tokens.RuleLoader, (c) => new RuleLoader(
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.IDataRegistry),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.RuleLoader}.`);

    // GenericContentLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, IDataRegistry, ILogger
    // NOTE: This might become deprecated or change with modding system.
    registrar.singletonFactory(tokens.GenericContentLoader, (c) => new GenericContentLoader(
        c.resolve(tokens.IConfiguration),
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.IDataRegistry),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.GenericContentLoader}.`);

    // ComponentDefinitionLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, IDataRegistry, ILogger
    registrar.singletonFactory(tokens.ComponentDefinitionLoader, (c) => new ComponentDefinitionLoader(
        c.resolve(tokens.IConfiguration),
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.IDataRegistry),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.ComponentDefinitionLoader}.`);

    // GameConfigLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, ILogger
    registrar.singletonFactory(tokens.GameConfigLoader, (c) => new GameConfigLoader({
        configuration: c.resolve(tokens.IConfiguration),
        pathResolver: c.resolve(tokens.IPathResolver),
        dataFetcher: c.resolve(tokens.IDataFetcher),
        schemaValidator: c.resolve(tokens.ISchemaValidator),
        logger: c.resolve(tokens.ILogger)
    }));
    logger.debug(`Loaders Registration: Registered ${tokens.GameConfigLoader}.`);

    // === ADDED: MODLOADER-005 A START ===
    // ModManifestLoader depends on IConfiguration, IPathResolver, IDataFetcher, ISchemaValidator, IDataRegistry, ILogger
    // NOTE: Assuming tokens.ModManifestLoader exists in src/core/tokens.js
    // If src/core/tokens.js was not provided, you would need to add:
    // ModManifestLoader: Symbol('ModManifestLoader')
    // to the tokens object definition in that file.
    registrar.singletonFactory(tokens.ModManifestLoader, c => new ModManifestLoader(
        c.resolve(tokens.IConfiguration),
        c.resolve(tokens.IPathResolver),
        c.resolve(tokens.IDataFetcher),
        c.resolve(tokens.ISchemaValidator),
        c.resolve(tokens.IDataRegistry),
        c.resolve(tokens.ILogger)
    ));
    logger.debug(`Loaders Registration: Registered ${tokens.ModManifestLoader}.`);
    // === ADDED: MODLOADER-005 A END ===


    logger.info('Loaders Registration: Completed.');
}