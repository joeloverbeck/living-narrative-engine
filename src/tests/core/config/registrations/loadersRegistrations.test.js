// src/tests/core/config/registrations/loadersRegistrations.test.js

// --- JSDoc Imports for Type Hinting ---
/** @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger */
/** @typedef {import('../../interfaces/coreServices.js').IConfiguration} IConfiguration */
/** @typedef {import('../../interfaces/coreServices.js').IPathResolver} IPathResolver */
/** @typedef {import('../../interfaces/coreServices.js').ISchemaValidator} ISchemaValidator */
/** @typedef {import('../../interfaces/coreServices.js').IDataRegistry} IDataRegistry */
/** @typedef {import('../../interfaces/coreServices.js').IDataFetcher} IDataFetcher */
/** @typedef {import('../../services/schemaLoader.js').default} SchemaLoader */ // Ensure path is correct
// Correct the SchemaLoader import path if it's different, e.g.:
// src/tests/core/config/registrations/loadersRegistrations.test.js

// --- JSDoc Imports ---
// ... (keep existing imports) ...
/** @typedef {import('../../../../core/services/schemaLoader.js').default} SchemaLoader */ // Corrected path example
/** @typedef {any} AppContainer */

// --- Jest Imports ---
import {describe, beforeEach, afterEach, it, expect, jest} from '@jest/globals';

// --- Class Under Test ---
import {registerLoaders} from '../../../../core/config/registrations/loadersRegistrations.js';

// --- Dependencies ---
import {tokens} from '../../../../core/tokens.js';
import SchemaLoader from '../../../../core/services/schemaLoader.js'; // Import actual class

// --- Mock Implementations (Core Services) ---
const mockLogger = {info: jest.fn(), warn: jest.fn(), error: jest.fn(), debug: jest.fn()};

const mockConfiguration = {
    getBaseDataPath: jest.fn().mockReturnValue('./data'),
    getSchemaFiles: jest.fn().mockReturnValue(['common.schema.json', 'entity.schema.json']),
    getContentTypeSchemaId: jest.fn((typeName) => { /* ... */
    }),
    getSchemaBasePath: jest.fn().mockReturnValue('./data/schemas'),
    getContentBasePath: jest.fn(typeName => `./data/${typeName}`),
    getWorldBasePath: jest.fn().mockReturnValue('./data/worlds'),
    getGameConfigFilename: jest.fn().mockReturnValue('game.json'),
    getRuleBasePath: jest.fn().mockReturnValue('./data/system-rules'),
    getRuleSchemaId: jest.fn().mockReturnValue('http://example.com/schemas/system-rule.schema.json'),
    // getManifestSchemaId: jest.fn().mockReturnValue(undefined), // REMOVED - No longer needed/checked
};

// --- UPDATED mockPathResolver (Provide resolveSchemaPath) ---
const mockPathResolver = {
    resolveSchemaPath: jest.fn(filename => `resolved/schemas/${filename}`), // Method used by SchemaLoader
    resolve: jest.fn(path => `resolved/${path}`) // Keep if needed by other loaders
};

// --- UPDATED mockSchemaValidator (Provide isSchemaLoaded) ---
const mockSchemaValidator = {
    validate: jest.fn().mockReturnValue({valid: true}), // Keep if needed by other loaders
    addSchema: jest.fn(), // Method used by SchemaLoader
    isSchemaLoaded: jest.fn().mockReturnValue(false) // Method used by SchemaLoader
};

// --- UPDATED mockDataFetcher (Provide fetch) ---
const mockDataFetcher = {
    fetch: jest.fn().mockResolvedValue({$id: 'http://example.com/schemas/common.schema.json'}), // Method used by SchemaLoader
    fetchText: jest.fn().mockResolvedValue('{}') // Keep if needed by other loaders
};

const mockDataRegistry = {register: jest.fn(), get: jest.fn()};


// --- Mock Custom DI Container ---
const createMockContainer = () => {
    const registrations = new Map();
    // --- Use a spy on the actual resolve implementation ---
    let containerInstance;
    const resolveSpy = jest.fn((token) => {
        const registration = registrations.get(token);
        if (!registration) {
            // Fallback logic
            if (token === tokens.ILogger) return mockLogger;
            if (token === tokens.IConfiguration) return mockConfiguration;
            if (token === tokens.IPathResolver) return mockPathResolver;
            if (token === tokens.ISchemaValidator) return mockSchemaValidator;
            if (token === tokens.IDataFetcher) return mockDataFetcher;
            if (token === tokens.IDataRegistry) return mockDataRegistry;
            throw new Error(`Mock Resolve Error: Token not registered or explicitly mocked: ${String(token)}`);
        }
        const {factoryOrValue, options} = registration;
        if (options?.lifecycle === 'singleton') {
            if (registration.instance === undefined) {
                registration.instance = typeof factoryOrValue === 'function' ? factoryOrValue(containerInstance) : factoryOrValue;
            }
            return registration.instance;
        }
        return typeof factoryOrValue === 'function' ? factoryOrValue(containerInstance) : factoryOrValue;
    });

    containerInstance = {
        _registrations: registrations,
        register: jest.fn((token, factoryOrValue, options = {}) => {
            if (!token) throw new Error('Mock Register Error: Token is required.');
            registrations.set(token, {factoryOrValue, options, instance: undefined});
        }),
        // Assign the spy to the container's resolve method
        resolve: resolveSpy,
    };
    return containerInstance;
};


describe('registerLoaders (with Mock DI Container)', () => {
    /** @type {ReturnType<typeof createMockContainer>} */
    let mockContainer;

    beforeEach(() => {
        jest.clearAllMocks(); // Clears mocks from previous tests
        mockContainer = createMockContainer(); // Creates a fresh container with fresh spies

        // Register the logger BEFORE calling the function under test
        // This call SHOULD be counted by the assertion in the test.
        mockContainer.register(tokens.ILogger, mockLogger, {lifecycle: 'singleton'});

        // Clear mocks for dependencies that might be called during registration
        mockPathResolver.resolve?.mockClear();
        mockPathResolver.resolveSchemaPath.mockClear();
        mockSchemaValidator.validate?.mockClear();
        mockSchemaValidator.addSchema.mockClear();
        mockSchemaValidator.isSchemaLoaded.mockClear();
        mockDataRegistry.register.mockClear();
        mockDataRegistry.get.mockClear();
        mockDataFetcher.fetch?.mockClear();
        mockDataFetcher.fetchText?.mockClear();
        Object.values(mockConfiguration).forEach(mockFn => typeof mockFn?.mockClear === 'function' && mockFn.mockClear());

        // Clear the container's RESOLVE spy, as its calls depend on the specific test case
        mockContainer.resolve.mockClear();
        // DO NOT clear mockContainer.register here, as we want to count the ILogger registration above
        // along with the registrations done by registerLoaders.
    });

    // Test 'should register all...' passes (assuming previous fixes are applied)
    it('should register all 11 services/loaders as singletons', () => {
        // Arrange: Logger is already registered in beforeEach

        // Act: Register the loaders
        registerLoaders(mockContainer);

        // Assert
        const expectedTokens = [
            tokens.IConfiguration, tokens.IPathResolver, tokens.ISchemaValidator,
            tokens.IDataRegistry, tokens.IDataFetcher, tokens.SchemaLoader,
            tokens.RuleLoader, tokens.GenericContentLoader,
            tokens.ComponentDefinitionLoader, tokens.GameConfigLoader, tokens.ModManifestLoader
        ];
        // Expect 1 (ILogger from beforeEach) + 11 (from registerLoaders) = 12 calls
        expect(mockContainer.register).toHaveBeenCalledTimes(expectedTokens.length + 1);

        // Check that each expected token was registered (ensures no duplicates or wrong tokens)
        expectedTokens.forEach(token => {
            expect(mockContainer.register).toHaveBeenCalledWith(
                token,           // The token itself
                expect.any(Function), // The factory function
                {lifecycle: 'singleton'} // The options object
            );
        });

        // Check the ILogger registration from beforeEach separately if needed, though the count implies it.
        expect(mockContainer.register).toHaveBeenCalledWith(
            tokens.ILogger,
            mockLogger, // The mock implementation itself (not a factory)
            {lifecycle: 'singleton'}
        );


        // Verify logger calls within registerLoaders
        // Note: The count might change slightly depending on whether the Registrar helper logs.
        // Assuming the pattern `logger.debug('Loaders Registration: Registered ${token}.')`
        expect(mockLogger.debug).toHaveBeenCalledTimes(1 + expectedTokens.length); // 1 "Starting..." + 11 "Registered..."
        expect(mockLogger.info).toHaveBeenCalledTimes(1); // 1 "Completed."
        expect(mockLogger.info).toHaveBeenCalledWith('Loaders Registration: Completed.');
    });


    it('should resolve SchemaLoader successfully (happy path) and respect singleton lifecycle', () => {
        // Arrange
        // Pre-register logger again for this specific test's container instance
        mockContainer.register(tokens.ILogger, mockLogger, {lifecycle: 'singleton'});
        // Call registerLoaders AFTER registering logger and BEFORE resolving anything else
        registerLoaders(mockContainer);

        // Act: Resolve SchemaLoader twice
        const loader1 = mockContainer.resolve(tokens.SchemaLoader);
        const loader2 = mockContainer.resolve(tokens.SchemaLoader);

        // Assert
        expect(loader1).toBeDefined();
        expect(loader1).toBeInstanceOf(SchemaLoader);
        expect(loader1).toBe(loader2); // Singleton check

        // Verify resolve was called *with* SchemaLoader token (twice explicitly)
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.SchemaLoader);

        expect(mockContainer.resolve).toHaveBeenCalledTimes(10);

        // Verify dependencies were resolved *by the factory* when resolve was called the first time.
        // These checks verify that the factory *did* ask for these tokens at least once.
        const callsToResolve = mockContainer.resolve.mock.calls;
        expect(callsToResolve).toEqual(
            expect.arrayContaining([
                [tokens.SchemaLoader], // From explicit call 1
                [tokens.SchemaLoader], // From explicit call 2
                [tokens.IConfiguration], // From SchemaLoader factory AND IPathResolver factory
                [tokens.IPathResolver], // From SchemaLoader factory
                [tokens.IDataFetcher], // From SchemaLoader factory
                [tokens.ISchemaValidator], // From SchemaLoader factory
                [tokens.ILogger] // From registerLoaders AND SchemaLoader factory
            ])
        );

        // Specific checks (still valid, check they were called at least once)
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.IConfiguration);
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.IPathResolver);
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.IDataFetcher);
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.ISchemaValidator);
        expect(mockContainer.resolve).toHaveBeenCalledWith(tokens.ILogger);
    });

    it('should match snapshot for registration calls', () => {
        registerLoaders(mockContainer);
        expect(mockContainer.register.mock.calls).toMatchSnapshot();
    });
});