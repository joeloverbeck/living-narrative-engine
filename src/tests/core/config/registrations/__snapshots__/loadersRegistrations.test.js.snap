// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerLoaders (with Mock DI Container) should match snapshot for registration calls 1`] = `
[
  [
    "ILogger",
    {
      "debug": [MockFunction] {
        "calls": [
          [
            "Loaders Registration: Starting core services and data loaders...",
          ],
          [
            "Loaders Registration: Registered IConfiguration.",
          ],
          [
            "Loaders Registration: Registered IPathResolver.",
          ],
          [
            "Loaders Registration: Registered ISchemaValidator.",
          ],
          [
            "Loaders Registration: Registered IDataRegistry.",
          ],
          [
            "Loaders Registration: Registered IDataFetcher.",
          ],
          [
            "Loaders Registration: Registered SchemaLoader.",
          ],
          [
            "Loaders Registration: Registered RuleLoader.",
          ],
          [
            "Loaders Registration: Registered ComponentDefinitionLoader.",
          ],
          [
            "Loaders Registration: Registered GameConfigLoader.",
          ],
          [
            "Loaders Registration: Registered ModManifestLoader.",
          ],
          [
            "Loaders Registration: Registered ActionLoader.",
          ],
          [
            "Loaders Registration: Registered EventLoader.",
          ],
          [
            "Loaders Registration: Registered EntityLoader.",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
          {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "error": [MockFunction],
      "info": [MockFunction] {
        "calls": [
          [
            "Loaders Registration: Completed.",
          ],
        ],
        "results": [
          {
            "type": "return",
            "value": undefined,
          },
        ],
      },
      "warn": [MockFunction],
    },
    {
      "isInstance": true,
      "lifecycle": "singleton",
    },
  ],
  [
    "IConfiguration",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "IPathResolver",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ISchemaValidator",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "IDataRegistry",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "IDataFetcher",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "SchemaLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "RuleLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ComponentDefinitionLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "GameConfigLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ModManifestLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "ActionLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "EventLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
  [
    "EntityLoader",
    [Function],
    {
      "lifecycle": "singleton",
    },
  ],
]
`;
